File: tests/interval.py
import math

def get_hack_interval(n, hack_time):
    """
    Return the hack interval for index n:
      H(n) = [hack_time/(2*n+1), hack_time/(2*n)]
    """
    lower = hack_time / (2 * n + 1)
    upper = hack_time / (2 * n)
    return (lower, upper)

def get_grow_interval(m, hack_time):
    """
    Return the grow interval for index m.
    Since grow_time = 3.2 * hack_time, we have:
      G(m) = [3.2*hack_time/(2*m+1), 3.2*hack_time/(2*m)]
    """
    grow_time = 3.2 * hack_time
    lower = grow_time / (2 * m + 1)
    upper = grow_time / (2 * m)
    return (lower, upper)

def intersect_intervals(interval1, interval2):
    """
    Given two intervals (a, b) and (c, d),
    return their intersection [max(a, c), min(b, d)] if nonempty;
    otherwise return None.
    """
    a, b = interval1
    c, d = interval2
    lower = max(a, c)
    upper = min(b, d)
    if lower <= upper:
        return (lower, upper)
    else:
        return None

def get_cropped_intervals(n, hack_time):
    """
    Given a hack interval H(n) for index n, find all grow intervals G(m)
    (with m chosen near 3.2*n) that intersect H(n).

    The method is:
      1. Set candidate m0 = round(3.2*n).
      2. Search downward (decreasing m) until G(m) no longer overlaps H(n).
      3. Search upward (increasing m) until G(m) no longer overlaps H(n).
      4. For each m in the found range, compute the intersection
         I = H(n) ∩ G(m).

    Returns a list of tuples (m, I) where I is the intersection interval.
    """
    # Get the hack interval for n
    H = get_hack_interval(n, hack_time)
    
    # Candidate m from the relation m ~ 3.2 * n.
    candidate_m = max(1, round(3.2 * n))
    
    # Search downward: find the smallest m for which G(m) still overlaps H.
    m_down = candidate_m
    while m_down >= 1:
        G = get_grow_interval(m_down, hack_time)
        if intersect_intervals(H, G) is None:
            break
        m_down -= 1
    m_min = m_down + 1  # last m that gave a nonempty intersection
    
    # Search upward: find the largest m for which G(m) overlaps H.
    m_up = candidate_m + 1
    while True:
        G = get_grow_interval(m_up, hack_time)
        if intersect_intervals(H, G) is None:
            break
        m_up += 1
    m_max = m_up - 1  # last m that worked
    
    # Collect the intersections for m in [m_min, m_max]
    intersections = []
    for m in range(m_min, m_max + 1):
        G = get_grow_interval(m, hack_time)
        inter = intersect_intervals(H, G)
        if inter is not None:
            intersections.append((m, inter))
    return intersections

# --- Example usage ---
hack_time = 1270.5873382302843  # example hack_time value
n = 1  # example hack interval index

H = get_hack_interval(n, hack_time)
print(f"Hack interval for n={n}: {H}")

cropped_intervals = get_cropped_intervals(n, hack_time)
print(f"\nFor hack interval n={n}, the grow intervals that overlap (and their intersections) are:")
for m, inter in cropped_intervals:
    G = get_grow_interval(m, hack_time)
    print(f"  For m={m}:")
    print(f"    Grow interval G({m}) = {G}")
    print(f"    Intersection with H({n}) = {inter}")

File: src/discarded/interval.ts
// Define an interval as a tuple [lower, upper]
type Interval = [number, number]

// Returns the hack interval for index n:
// H(n) = [hackTime/(2*n+1), hackTime/(2*n)]
function getHackInterval(n: number, hackTime: number): Interval {
  const lower = hackTime / (2 * n + 1)
  const upper = hackTime / (2 * n)
  return [lower, upper]
}

// Returns the grow interval for index m.
// Given that growTime = 3.2 * hackTime, we have:
// G(m) = [3.2*hackTime/(2*m+1), 3.2*hackTime/(2*m)]
function getGrowInterval(m: number, hackTime: number): Interval {
  const growTime = 3.2 * hackTime
  const lower = growTime / (2 * m + 1)
  const upper = growTime / (2 * m)
  return [lower, upper]
}

// Computes the intersection of two intervals.
// If the intervals [a, b] and [c, d] overlap, returns [max(a, c), min(b, d)].
// Otherwise, returns null.
function intersectIntervals(
  interval1: Interval,
  interval2: Interval
): Interval | null {
  const [a, b] = interval1
  const [c, d] = interval2
  const lower = Math.max(a, c)
  const upper = Math.min(b, d)
  return lower <= upper ? [lower, upper] : null
}

// An interface for the raw intersection result for a given m.
interface CroppedInterval {
  m: number
  intersection: Interval
}

/**
 * For a given hack interval H(n) (with index n), this function finds all grow intervals
 * G(m) that overlap H(n) by scanning downward and upward from an estimated candidate m.
 * It returns an array of raw intersections.
 */
function getCroppedIntervals(n: number, hackTime: number): CroppedInterval[] {
  const H = getHackInterval(n, hackTime)

  // Start with a candidate m ~ 3.2 * n.
  let candidateM = Math.max(1, Math.round(3.2 * n))

  // Search downward: decrease m until the grow interval no longer overlaps H.
  let mDown = candidateM
  while (mDown >= 1) {
    const G = getGrowInterval(mDown, hackTime)
    if (intersectIntervals(H, G) === null) {
      break
    }
    mDown--
  }
  const mMin = mDown + 1 // last m that produced an intersection

  // Search upward: increase m until the grow interval no longer overlaps H.
  let mUp = candidateM + 1
  while (true) {
    const G = getGrowInterval(mUp, hackTime)
    if (intersectIntervals(H, G) === null) {
      break
    }
    mUp++
  }
  const mMax = mUp - 1 // last m that produced an intersection

  // Collect all intersections for m in [mMin, mMax]
  const intersections: CroppedInterval[] = []
  for (let m = mMin; m <= mMax; m++) {
    const G = getGrowInterval(m, hackTime)
    const inter = intersectIntervals(H, G)
    if (inter !== null) {
      intersections.push({ m, intersection: inter })
    }
  }
  return intersections
}

/**
 * Merges an array of intervals into a union.
 * If intervals overlap or touch, they are merged into one.
 */
function mergeIntervals(intervals: Interval[]): Interval[] {
  if (intervals.length === 0) return []

  // Sort intervals by their lower bound.
  intervals.sort((a, b) => a[0] - b[0])

  const merged: Interval[] = []
  let current = intervals[0]

  for (let i = 1; i < intervals.length; i++) {
    const next = intervals[i]
    // If the current interval overlaps or touches the next interval, merge them.
    if (current[1] >= next[0]) {
      current = [current[0], Math.max(current[1], next[1])]
    } else {
      merged.push(current)
      current = next
    }
  }
  merged.push(current)
  return merged
}

/**
 * Given a hack interval H(n) for a specified index n, this function returns the final
 * union of intersections between H(n) and all overlapping grow intervals.
 * The result is an array of one or more intervals.
 */
function getFinalIntersections(n: number, hackTime: number): Interval[] {
  const rawCropped = getCroppedIntervals(n, hackTime)
  const intersections = rawCropped.map((ci) => ci.intersection)
  const merged = mergeIntervals(intersections)
  return merged
}

// --- New Example ---
// For each hack interval n from 1 to 10, compute the final union of intersections.
// If the union contains 2 or 3 intervals (groups), then print just their sizes (in order).

const hackTime: number = 1270.5873382302843 // example hackTime value

for (let n = 1; n <= 100; n++) {
  const finalIntersections = getFinalIntersections(n, hackTime)

  // Check if the final union has 2 or 3 intervals.
  if (finalIntersections.length === 2 || finalIntersections.length === 3) {
    // Compute the size (width) of each interval.
    const sizes = finalIntersections.map(
      (interval) => interval[1] - interval[0]
    )
    // Print the sizes in order.
    console.log(
      `n = ${n}: Interval sizes = ${sizes.map((s) => s.toFixed(2)).join(", ")} ms`
    )
  }
}
